File: clients/python_pypy/model/order.py
from model.unit_order import UnitOrder
from stream_wrapper import StreamWrapper
from typing import Dict

class Order:
    """Player's (team's) orders"""

    __slots__ = ("unit_orders",)

    unit_orders: Dict[int, UnitOrder]

    def __init__(self, unit_orders: Dict[int, UnitOrder]):
        self.unit_orders = unit_orders
        """Orders for each of your units"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Order":
        """Read Order from input stream
        """
        unit_orders = {}
        for _ in range(stream.read_int()):
            unit_orders_key = stream.read_int()
            unit_orders_value = UnitOrder.read_from(stream)
            unit_orders[unit_orders_key] = unit_orders_value
        return Order(unit_orders)
    
    def write_to(self, stream: StreamWrapper):
        """Write Order to output stream
        """
        stream.write_int(len(self.unit_orders))
        for key, value in self.unit_orders.items():
            stream.write_int(key)
            value.write_to(stream)
    
    def __repr__(self):
        return "Order(" + \
            repr(self.unit_orders) + \
            ")"

File: clients/cpp/model/Order.cpp
#include "Order.hpp"

namespace model {

Order::Order(std::unordered_map<int, model::UnitOrder> unitOrders) : unitOrders(unitOrders) { }

// Read Order from input stream
Order Order::readFrom(InputStream& stream) {
    size_t unitOrdersSize = stream.readInt();
    std::unordered_map<int, model::UnitOrder> unitOrders = std::unordered_map<int, model::UnitOrder>();
    unitOrders.reserve(unitOrdersSize);
    for (size_t unitOrdersIndex = 0; unitOrdersIndex < unitOrdersSize; unitOrdersIndex++) {
        int unitOrdersKey = stream.readInt();
        model::UnitOrder unitOrdersValue = model::UnitOrder::readFrom(stream);
        unitOrders.emplace(std::make_pair(unitOrdersKey, unitOrdersValue));
    }
    return Order(unitOrders);
}

// Write Order to output stream
void Order::writeTo(OutputStream& stream) const {
    stream.write((int)(unitOrders.size()));
    for (const auto& unitOrdersEntry : unitOrders) {
        const int& unitOrdersKey = unitOrdersEntry.first;
        const model::UnitOrder& unitOrdersValue = unitOrdersEntry.second;
        stream.write(unitOrdersKey);
        unitOrdersValue.writeTo(stream);
    }
}

// Get string representation of Order
std::string Order::toString() const {
    std::stringstream ss;
    ss << "Order { ";
    ss << "unitOrders: ";
    ss << "{ ";
    size_t unitOrdersIndex = 0;
    for (const auto& unitOrdersEntry : unitOrders) {
        if (unitOrdersIndex != 0) {
            ss << ", ";
        }
        const int& unitOrdersKey = unitOrdersEntry.first;
        const model::UnitOrder& unitOrdersValue = unitOrdersEntry.second;
        ss << unitOrdersKey;
        ss << ": ";
        ss << unitOrdersValue.toString();
        unitOrdersIndex++;
    }
    ss << " }";
    ss << " }";
    return ss.str();
}

}